File: src/test/java/at/ac/uibk/dps/cirrina/execution/service/HttpServiceImplementationTest.java
package at.ac.uibk.dps.cirrina.execution.service;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import at.ac.uibk.dps.cirrina.csm.Csml.ContextVariableDescription;
import at.ac.uibk.dps.cirrina.csm.ServiceImplementationBindings.HttpMethod;
import at.ac.uibk.dps.cirrina.execution.object.context.ContextVariable;
import at.ac.uibk.dps.cirrina.execution.object.context.ContextVariableBuilder;
import at.ac.uibk.dps.cirrina.execution.object.context.Extent;
import at.ac.uibk.dps.cirrina.execution.object.exchange.ContextVariableExchange;
import at.ac.uibk.dps.cirrina.execution.object.exchange.ContextVariableProtos;
import at.ac.uibk.dps.cirrina.execution.service.HttpServiceImplementation.Parameters;
import com.sun.net.httpserver.HttpServer;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.stream.Stream;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

public class HttpServiceImplementationTest {

  private static HttpServer httpServer;

  @BeforeAll
  public static void setUp() throws IOException {
    httpServer = HttpServer.create(new InetSocketAddress(8000), 0);

    httpServer.createContext("/plus", exchange -> {
      Assertions.assertEquals(
        "some-id",
        exchange.getRequestHeaders().get("Cirrina-Sender-ID").getFirst()
      );

      final var payload = exchange.getRequestBody().readAllBytes();

      final var in = ContextVariableProtos.ContextVariables.parseFrom(payload)
        .getDataList()
        .stream()
        .map(ContextVariableExchange::fromProto)
        .toList();

      final var varOne = in
        .stream()
        .filter(e -> e.name().equals("varOne"))
        .findFirst();
      final var varTwo = in
        .stream()
        .filter(e -> e.name().equals("varTwo"))
        .findFirst();

      // Create output
      final var out = ContextVariableProtos.ContextVariables.newBuilder()
        .addAllData(
          Stream.of(
            new ContextVariable("result", (int) varOne.get().value() + (int) varTwo.get().value())
          )
            .map(contextVariable -> new ContextVariableExchange(contextVariable).toProto())
            .toList()
        )
        .build()
        .toByteArray();

      // Response status and length
      exchange.sendResponseHeaders(200, out.length);

      // Output the response
      try (final var stream = exchange.getResponseBody()) {
        stream.write(out);
      }
    });

    httpServer.createContext("/error", exchange -> {
      // Response status and length
      exchange.sendResponseHeaders(500, 0);

      // Output the response
      try (final var stream = exchange.getResponseBody()) {
        stream.write(new byte[0]);
      }
    });

    httpServer.createContext("/broken-response1", exchange -> {
      // Response status and length
      exchange.sendResponseHeaders(200, 1);

      // Output the response
      try (final var stream = exchange.getResponseBody()) {
        stream.write(new byte[] { 1 });
      }
    });

    httpServer.createContext("/broken-response2", exchange -> {
      final byte[] out = null;

      // Response status and length
      exchange.sendResponseHeaders(200, out.length);

      // Output the response
      try (final var stream = exchange.getResponseBody()) {
        stream.write(out);
      }
    });

    httpServer.start();
  }

  @AfterAll
  public static void tearDown() {
    httpServer.stop(0);
  }

  @Test
  public void testHttpServiceInvocation() {
    List.of(HttpMethod.POST, HttpMethod.GET)
      .stream()
      .forEach(method -> {
        // First variable
        final var varOne = new ContextVariableDescription("varOne", "5");

        // Second variable
        final var varTwo = new ContextVariableDescription("varTwo", "6");

        // Success
        assertDoesNotThrow(() -> {
          final var variables = new ArrayList<ContextVariable>();
          variables.add(ContextVariableBuilder.from(varOne).build().evaluate(new Extent()));
          variables.add(ContextVariableBuilder.from(varTwo).build().evaluate(new Extent()));

          final var service = new HttpServiceImplementation(
            new Parameters("http", false, "http", "localhost", 8000, "/plus", HttpMethod.POST)
          );

          final var output = service.invoke(variables, "some-id").get();

          assertEquals(1, output.size());

          final var result = output.getFirst();
          assertEquals("result", result.name());
          assertEquals(11, result.value());
        });

        // HTTP error
        assertThrows(ExecutionException.class, () -> {
          final var service = new HttpServiceImplementation(
            new Parameters("http", false, "http", "localhost", 8000, "/error", HttpMethod.POST)
          );

          service.invoke(new ArrayList<>(), "some-id").get();
        });

        // Invalid response type
        assertThrows(ExecutionException.class, () -> {
          final var service = new HttpServiceImplementation(
            new Parameters(
              "http",
              false,
              "http",
              "localhost",
              8000,
              "/broken-response1",
              HttpMethod.POST
            )
          );

          service.invoke(new ArrayList<>(), "some-id").get();
        });

        // Invalid response type
        assertThrows(ExecutionException.class, () -> {
          final var service = new HttpServiceImplementation(
            new Parameters(
              "http",
              false,
              "http",
              "localhost",
              8000,
              "/broken-response2",
              HttpMethod.POST
            )
          );

          service.invoke(new ArrayList<>(), "some-id").get();
        });
      });
  }
}

File: src/test/java/at/ac/uibk/dps/cirrina/execution/service/HttpServiceImplementationTest.kt
package at.ac.uibk.dps.cirrina.execution.service

import at.ac.uibk.dps.cirrina.csm.ServiceImplementationBindings.HttpMethod
import at.ac.uibk.dps.cirrina.execution.`object`.context.ContextVariable
import at.ac.uibk.dps.cirrina.execution.`object`.context.ContextVariableBuilder
import at.ac.uibk.dps.cirrina.execution.`object`.context.Extent
import at.ac.uibk.dps.cirrina.execution.`object`.exchange.ContextVariableExchange
import at.ac.uibk.dps.cirrina.execution.`object`.exchange.ContextVariableProtos
import com.sun.net.httpserver.HttpExchange
import com.sun.net.httpserver.HttpServer
import java.net.InetSocketAddress
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.*
import org.junit.jupiter.api.Assertions.assertEquals

class HttpServiceImplementationTest {

  companion object {
    private lateinit var httpServer: HttpServer

    @JvmStatic
    @BeforeAll
    fun setUp() {
      httpServer =
        HttpServer.create(InetSocketAddress(8000), 0).apply {
          createContext("/plus") { exchange ->
            exchange
              .readVariables()
              .let { vars ->
                val v1 = vars.first { it.name == "varOne" }.value as Int
                val v2 = vars.first { it.name == "varTwo" }.value as Int
                listOf(ContextVariable("result", v1 + v2))
              }
              .let { result -> exchange.sendVariables(result) }
          }

          createContext("/error") { it.sendResponseHeaders(500, -1) }

          createContext("/broken-response1") { exchange ->
            exchange.sendResponseHeaders(200, 1)
            exchange.responseBody.use { it.write(byteArrayOf(1)) }
          }

          start()
        }
    }

    @JvmStatic @AfterAll fun tearDown() = httpServer.stop(0)

    private fun HttpExchange.readVariables(): List<ContextVariable> =
      requestBody.use { stream ->
        ContextVariableProtos.ContextVariables.parseFrom(stream.readAllBytes())
          .dataList
          .map(ContextVariableExchange::fromProto)
      }

    private fun HttpExchange.sendVariables(variables: List<ContextVariable>) =
      variables
        .map { ContextVariableExchange(it).toProto() }
        .let { protos ->
          ContextVariableProtos.ContextVariables.newBuilder()
            .addAllData(protos)
            .build()
            .toByteArray()
        }
        .let { bytes ->
          sendResponseHeaders(200, bytes.size.toLong())
          responseBody.use { it.write(bytes) }
        }
  }

  @Test
  fun testHttpServiceInvocation() = runBlocking {
    listOf(HttpMethod.POST, HttpMethod.GET).forEach { method ->
      // Success case
      assertDoesNotThrow {
          createService("/plus", method)
            .invoke(
              listOf(
                ContextVariableBuilder.empty()
                  .name("varOne")
                  .value(5)
                  .build()
                  .getOrThrow()
                  .evaluate(Extent.empty()),
                ContextVariableBuilder.empty()
                  .name("varTwo")
                  .value(6)
                  .build()
                  .getOrThrow()
                  .evaluate(Extent.empty()),
              )
            )
        }
        .let { result ->
          assertEquals(1, result.size)
          assertEquals("result", result.first().name)
          assertEquals(11, result.first().value)
        }

      // Error case
      assertThrows<IllegalStateException> { createService("/error", method).invoke(emptyList()) }

      // Broken serialization case
      assertThrows<IllegalStateException> {
        createService("/broken-response1", method).invoke(emptyList())
      }
    }
  }

  private fun createService(path: String, method: HttpMethod) =
    HttpServiceImplementation("http", "localhost", 8000, path, method, "test-service", false)
}
